pkg/worker/worker.go

```diff
@@ -0,0 +1,278 @@
+// Job is an abstraction around Linux process that contains logs and metadata.
+type Job struct {
+	DataStorage DataStorage
+	Command     []string
+	Status      JobStatus
+	PID         int
+	UUID        googleUUID.UUID
+	CreatedAt   time.Time
+	StartedAt   time.Time
+	StoppedAt   time.Time
+	FinishedAt  time.Time
+	User        string
+	Logs        []string
+}
+
+// newJob assigns the default, required values returning new job instance.
+func newJob(dataStorage DataStorage, command []string, user string) Job {
+	return Job{
+		DataStorage: dataStorage,
+		Status:      StatusCreated,
+		Command:     command,
+		UUID:        googleUUID.New(),
+		CreatedAt:   time.Now(),
+		User:        user,
+		Logs:        []string{},
+	}
+}
+
+// start runs the the command in a new process, listens for StdIn and StdErr
+// and logs them to the DataStorage.
+// If the DataStorage is not accessible, the errors are raised up to the Worker.Start
+// method and logged to the standard output.
+// TODO: Add stopping subprocesses of the started processes.
+// TODO: Add separation of the StdIn and StdErr.
+func (job *Job) start() error {
+	cmd := exec.Command(job.Command[0], job.Command[1:]...)
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setsid: true}
+
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		err = errors.Wrap(err, "error creating StdoutPipe for Cmd")
+		return job.logAndReturnError(err)
+	}
+	cmd.Stderr = cmd.Stdout
+
+	err = cmd.Start()
+	if err != nil {
+		err = errors.Wrap(err, "error starting process")
+		return job.logAndReturnError(err)
+	}
+
+	job.Status = StatusRunning
+	job.PID = cmd.Process.Pid
+	job.StartedAt = time.Now()
+	err = job.store()
+	if err != nil {
+		return err
+	}
+
+	buffer := make([]byte, 1000)
+	for {
+		n, err := stdout.Read(buffer)
+		if err == nil {
+			job.Logs = append(job.Logs, string(buffer[:n]))
+			err = job.store()
+			if err != nil {
+				return err
+			}
+		}
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return errors.Wrap(err, "error storing command ouput in logs")
+		}
+	}
+
+	err = cmd.Wait()
+
+	if err != nil {
+		err = errors.Wrap(err, "error during process execution")
+		return job.logAndReturnError(err)
+	}
+
+	job.FinishedAt = time.Now()
+	job.Status = StatusFinished
+	err = job.store()
+
+	return err
+}
+
+// store saves the job object to DataStorage.
+func (job *Job) store() error {
+	err := job.DataStorage.Store(*job)
+	if err != nil {
+		return errors.Wrap(err, "error invoking Store method of the DataStorage")
+	}
+	return nil
+}
+
+// logAndReturnError tries to save the error to the logs of the job. If it fails,
+// the method wraps the error with the DataStorage error and returns it.
+func (job *Job) logAndReturnError(systemErr error) error {
+	job.Status = StatusError
+	job.Logs = append(job.Logs, systemErr.Error())
+	err := job.store()
+	if err != nil {
+		return errors.Wrap(err, systemErr.Error())
+	}
+
+	return systemErr
+}
+
+func (job *Job) wait() {
+	time.Sleep(1 * time.Millisecond)
+}
```

Comment:
sorry, I meant that `job.wait` should actually wait for a job to end (get into the `stopped`, `error` or `finished` state), not just sleep